Add basic Aranet integration

aschmitz opened the pull request:
## Proposed change
This PR adds support for the [Aranet4](https://aranet.com/products/aranet4/) environmental (CO2, temperature, humidity, pressure) sensor.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24674
- This uses the third-party `aranet4` Python module available at https://github.com/Anrijs/Aranet4-Python.
- The code is heavily based on other BLE-using devices' integrations, especially those using [Bluetooth-Devices](https://github.com/Bluetooth-Devices), and in particular the `qingping` integration, but the upstream device library is not part of the Bluetooth-Devices collection.

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. (Not applicable for a new dependency?)
- [ ] Untested files have been added to `.coveragerc`. (N/A, I believe.)

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

aschmitz committed:
Add basic Aranet4 integration

bdraco commented on homeassistant/components/aranet4/manifest.json:
```diff
@@ -0,0 +1,26 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/aranet4",
+  "requirements": ["aranet4==2.1.3"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
```suggestion
```

aschmitz replied:
Ah, didn't realize I could leave these out. Thanks!

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+            if address in current_addresses or address in self._discovered_devices:
+                continue
+
+            # The aranet4 library expects the bleak format of objects, so if
+            # they haven't been passed through from hass (often they are), we
+            # generate them out of the discovery info.
+            if hasattr(discovery_info, "device"):
+                device = discovery_info.device
+            else:
+                device = BLEDevice(discovery_info.address, name=discovery_info.name)
+
+            if hasattr(discovery_info, "advertisement"):
+                ad_data = discovery_info.advertisement
+            else:
+                ad_data = AdvertisementData(
+                    local_name=None,
+                    manufacturer_data=discovery_info.manufacturer_data,
+                    service_data=discovery_info.service_data,
+                    service_uuids=discovery_info.service_uuids,
+                    rssi=discovery_info.rssi,
+                    tx_power=-127,
+                    platform_data=(),
+                )
```
`async_discovered_service_info` returns `BluetoothServiceInfoBleak` objects.

You can grab these off of them instead
https://github.com/home-assistant/core/blob/6e19629539d0dfecd7500e84ff067b2f6b856770/homeassistant/components/bluetooth/models.py#L49

aschmitz replied:
Handy, thanks. I've switched to that and updated the relevant test init to build `BluetoothServiceInfoBleak` objects, which cleans up this code.

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    def __init__(self) -> None:
+        """Set up a new config flow for Aranet4."""
+        self._discovered_devices: dict[str, tuple[str, Aranet4Advertisement]] = {}
+
```
You have `bluetooth` set in the `manifest.json` so you should implement `async_step_bluetooth`

Example:
https://github.com/home-assistant/core/blob/6e19629539d0dfecd7500e84ff067b2f6b856770/homeassistant/components/govee_ble/config_flow.py#L31

aschmitz replied:
I had vaguely been hoping to defer digging into that for a future PR, but I've added it now and tested it (both via automatic tests and manually).

bdraco commented on homeassistant/components/aranet4/sensor.py:
```diff
@@ -0,0 +1,159 @@
+    ),
+}
+
+SENSOR_NAMES = {
+    "temperature": "Temperature",
+    "humidity": "Humidity",
+    "pressure": "Pressure",
+    "co2": "Carbon Dioxide",
+    "battery": "Battery",
+    "interval": "Update Interval",
+}
```
It would be a bit cleaner to use the `name` field in the `SensorEntityDescription`

aschmitz replied:
Sure, might as well. It seems I still have to set `entity_names` in the `PassiveBluetoothDataUpdate`, but at least there's only one dict with all of these keys.

aschmitz committed:
fix: don't rebuild Bleak objects in config_flow

Pre review comment, this should be reliable inside hass, so we build
those in the test instead.

aschmitz committed:
fix: remove unnecessary lines from manifest.json

aschmitz committed:
chore: clean up sensor names in aranet4

It seems these still must be set in entity_names, but this helps keep
everything together.

aschmitz committed:
feat: support automatic Bluetooth config

aschmitz committed:
fix: handle loss of aranet4 smart home integration

aschmitz commented:
Thanks for the review, @bdraco! I've made the changes you pointed out, and done some additional testing with disabling the smart home integration of the Aranet4 after it was configured, leading to better support for "unavailable" states.

A question since you seem to be familiar with the internals of various Bluetooth integration bits: the device supports some configuration and a bit of additional information (such as serial number and hardware revision) via an active connection rather than a passive one: are you aware of any integrations that combine connectable and passive modes for the same Bluetooth device in the same integration? I'm not sure how to handle things like "how do you handle a switch entity that only works if the device is available via a connectable Bluetooth integration?".

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+    async_discovered_service_info,
+)
+from homeassistant.const import CONF_ADDRESS
+from homeassistant.data_entry_flow import FlowResult
```
```suggestion
from homeassistant.data_entry_flow import AbortFlow, FlowResult
```

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+
+        # If integrations are disabled, we get no sensor data.
+        if not adv.manufacturer_data.integrations:
+            return self.async_abort(reason="integrations_disabled")
```
```suggestion
            raise AbortFlow("integrations_disabled")
```

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        """Get any configuration errors that apply to an advertisement."""
+        # Old versions of firmware don't expose sensor data in advertisements.
+        if not adv.manufacturer_data or adv.manufacturer_data.version < MIN_VERSION:
+            return self.async_abort(reason="outdated_version")
```
```suggestion
            raise AbortFlow("outdated_version")
```

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        self._discovered_device: Aranet4Advertisement | None = None
+        self._discovered_devices: dict[str, tuple[str, Aranet4Advertisement]] = {}
+
+    def _errors_from_adv(self, adv: Aranet4Advertisement) -> FlowResult | None:
```
```suggestion
    def _errors_from_adv(self, adv: Aranet4Advertisement) -> None:
```

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        if not adv.manufacturer_data.integrations:
+            return self.async_abort(reason="integrations_disabled")
+
+        return None
```
```suggestion
```

bdraco commented on homeassistant/components/aranet4/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        adv = Aranet4Advertisement(discovery_info.device, discovery_info.advertisement)
+        errors = self._errors_from_adv(adv)
+        if errors:
+            return errors
```
```suggestion
        self._errors_from_adv(adv)
```

Probably rename this to `_raise_for_advertisement_errors`

bdraco commented:
> A question since you seem to be familiar with the internals of various Bluetooth integration bits: the device supports some configuration and a bit of additional information (such as serial number and hardware revision) via an active connection rather than a passive one: are you aware of any integrations that combine connectable and passive modes for the same Bluetooth device in the same integration? I'm not sure how to handle things like "how do you handle a switch entity that only works if the device is available via a connectable Bluetooth integration?".

In a future PR you can switch to using the `ActiveBluetoothProcessorCoordinator`

Examples:
https://github.com/home-assistant/core/blob/d5a46c3a1068ec22dd2b0181a6af690668145f4b/homeassistant/components/xiaomi_ble/__init__.py#L86
https://github.com/home-assistant/core/blob/d5a46c3a1068ec22dd2b0181a6af690668145f4b/homeassistant/components/xiaomi_ble/__init__.py#L65

aschmitz committed:
chore: clean up abort logic in aranet4 config_flow

aschmitz commented:
Thanks, that does clean things up. I'll probably look into `ActiveBluetoothProcessorCoordinator` in the future too, though I suspect the passive integration will get folks 90%+ of the utility.

bdraco commented:
coverage is good

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform darwin, python 3.10.8-final-0 ----------
Name                                              Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------
homeassistant/components/aranet4/__init__.py         25      0   100%
homeassistant/components/aranet4/config_flow.py      62      0   100%
homeassistant/components/aranet4/const.py             1      0   100%
homeassistant/components/aranet4/sensor.py           36      0   100%
-------------------------------------------------------------------------------
TOTAL                                               124      0   100%
```

bdraco commented:
I think this is good to go but we do need to rename it to `aranet` since we won't want an `aranet5`, `aranet6`, etc integration when they release a new device model.

aschmitz commented:
> I think this is good to go but we do need to rename it to `aranet` since we won't want an `aranet5`, `aranet6`, etc integration when they release a new device model.

I'm fine with that if we want, but the Aranet4 _appears_ to refer to the four sensors it has (CO2, temperature, humidity, pressure) rather than anything else: there was no Aranet2 or Aranet3, etc. (There are _other_ Aranet products, but those appear to be aimed at industrial customers and a different ecosystem that doesn't seem to have Bluetooth support.)

aschmitz commented:
(If you do indeed still want it to be `aranet` instead, let me know and I'll update the PRs.)

bdraco commented:
> (If you do indeed still want it to be `aranet` instead, let me know and I'll update the PRs.)

Yes please

aschmitz committed:
chore: rename `aranet4` integration to `aranet`

bdraco commented:
> Okay, I believe that's done. The integration is still _named_ Aranet4 in the interface, but the domain is `aranet` rather than `aranet4`, so it's an easy change if another device comes out. (My only worry with naming it "Aranet" in the UI / docs is that it's not actually compatible with other [Aranet sensors](https://aranet.com/products/), so I don't want to be misleading.)

Usually we name it for the brand/vendor and deal with that by listing which devices are supported example:

https://www.home-assistant.io/integrations/qingping/#supported-devices

aschmitz committed:
chore: complete renaming Aranet4 to Aranet

aschmitz commented:
> Usually we name it for the brand/vendor and deal with that by listing which devices are supported example:

Alrighty, that should be done too.

balloob commented on homeassistant/components/aranet/sensor.py:
```diff
@@ -0,0 +1,169 @@
+        device_class=SensorDeviceClass.DURATION,
+        native_unit_of_measurement=TIME_SECONDS,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
```
This entity should have been disabled by default as it's not useful for most people.

```
        entity_registry_enabled_default=False,
```

@aschmitz would you be able to PR that change?
